### PR TITLE
DOC: drastic rewrite of constants documentation

### LINKED ISSUE #11173: DOC: Major overhaul of scipy.constants documentation?
Hi,

I personally find the [scipy.constants](https://docs.scipy.org/doc/scipy/reference/constants.html) a not very concise collection of units and constants.
If you don't have any objections, I would write up a draft following the SI-brochure. This could either mean a restructured and extended documentation leaving all the existing information on the page or a slimmed down variant that drops the units that are not part of the SI conventions.

Regards

### PR DESCRIPTION
**Pull request by jakobjakobson13**
If this rewrite is too drastic, let me know.

#### Reference issue
Fixes #11173 

#### What does this implement/fix?
It is an overhaul of the constants documentation.

#### Related pull requests
#11343 and #11345 are also concerning the constants module however with a different focus.

### CODE REVIEW

**Comment by ilayn** [comment]
@jakobjakobson13 Thanks a lot for doing this. Is it possible that you work on this file offline such that it doesn't choke the CI/CD pipeline? Because you are placing a commit each time you make a change it is firing up a new trigger for all the tools as you can see below.

**Comment by jakobjakobson13** [comment]
> @jakobjakobson13 Thanks a lot for doing this. Is it possible that you work on this file offline such that it doesn't choke the CI/CD pipeline? Because you are placing a commit each time you make a change it is firing up a new trigger for all the tools as you can see below.

I did not manage to compile the documentation at my laptop but I'll try it for the next commits. For the moment I would wait anyway for any comments of you guys before I push/make any further commits.

File: scipy/constants/__init__.py [hunk 1]
```diff
@@ -5,316 +5,104 @@
...
------
-
-====================  =======================================================
-``hp``                one horsepower in watts
-``horsepower``        one horsepower in watts
-====================  =======================================================
-
-Force
------
+or via the dictionaries stored in :mod:`scipy.constants.physical_constants`
```
Comment (rgommers) [inline scipy/constants/__init__.py]: This is the cause of the doc build warnings, there is no module `physical_constants`. Shouldn't this just be `scipy.constants`?

**Comment by rgommers** [comment]
You can see the rendered version in the `build_docs_artifact` CI link. There's still some issues with links.

I like the restructure, first a sensible explanation and the huge list at the end makes sense.

File: scipy/constants/__init__.py [hunk 2]
```diff
@@ -5,316 +5,104 @@
...
-``golden_ratio``  Golden ratio
-================  =================================================================
-
-
-Physical constants
+About the SI units
 ==================
+In 2019 seven système international d’unités (SI) base units were (re)defined to
+serve as a basis for all physical quantities by the bureau
+international des poids et mesures (BIPM) [SIunits]_. They can be summarized as followed
```
Comment (rgommers) [inline scipy/constants/__init__.py]: the `[ ]` brackets need removing.

File: scipy/constants/__init__.py [hunk 3]
```diff
@@ -5,316 +5,104 @@
...
-``long_ton``       one long ton in kg
-``short_ton``      one short ton in kg
-``troy_ounce``     one Troy ounce in kg
-``troy_pound``     one Troy pound in kg
-``carat``          one carat in kg
-``m_u``            atomic mass constant (in kg)
-``u``              atomic mass constant (in kg)
-``atomic_mass``    atomic mass constant (in kg)
-=================  ============================================================
+- The ampere :math:`\mathrm{A}` as the SI unit of electric current is derived
```
Comment (rgommers) [inline scipy/constants/__init__.py]: This would read better as `ampere (A)` rather than `ampere A`. And same for next lines.

File: scipy/constants/__init__.py (hunk 1 above)
Comment (rgommers) [reply scipy/constants/__init__.py]: Ah no, you are wanting to refer to `physical_constants`, which is simply a dictionary not a module (so both the `:mod:` is wrong and the sentence may need a tweak:
```
In [3]: type(constants.physical_constants)                                                     
Out[3]: dict
```

File: scipy/constants/__init__.py (hunk 2 above)
Comment (jakobjakobson13) [reply scipy/constants/__init__.py]: Well, here I wanted to provide the link with background information. Should I delete it anyway?

Comment (rgommers) [reply scipy/constants/__init__.py]: No the link is fine, it's just the brackets that are now part of the link and look weird, also `SIunits` isn't the most informative name. How about simply:
```
(see [1])
```

**Comment by jakobjakobson13** [comment]
Okay, so CircleCI fails due to `release/0.18.0-notes.rst:215: WARNING: py:obj reference target not found: scipy.constants.convert_temperature` and TravisCI fails due to `ERROR: objects in scipy.constants.__all__ but not in refguide::`. The other failures are, as far as I understand, however unrelated to my pull request.

So, how should this be handled?

**Comment by rgommers** [comment]
The `convert_temperature` one is straightforward, just putting double backticks around that in the 0.18.0 release notes.

For the refguide check, one possible fix is to add all those objects to the skiplist at https://github.com/scipy/scipy/blob/master/tools/refguide_check.py#L127 (in an efficient way, using a list comprehension and a string join, rather than repeating `r'scipy\.constants\.` over and over).

The argument for the skiplist fix would be that constants don't need separate pages - they don't have docstrings. _However_, that runs the risk of the list being/becoming incomplete. So probably the more robust fix is to make sure the long listing you have is of the right form (typeset as links rather than code). I'll comment on the diff for the fix.

**Comment by rgommers** [comment]
This looks a little weird: 

image

the first three lines shouldn't have line breaks between them I think.

File: scipy/constants/__init__.py [hunk 4]
```diff
@@ -5,316 +5,107 @@
...
+names and serveral non-SI units. Of course all of before mentioned units
+can be used with the SI prefixes.
+
+However, one has to keep in mind that different systems of measurements apply
+for different problems and therefore the SI units are not necessarily the
+best (in terms of convenience) units for every problem. An overview of
+different systems of measurement can be found on Wikipedia [Wikipedia]_ .
 
-===========================  =================================================================
-``c``                        speed of light in vacuum
```
Comment (rgommers) [inline scipy/constants/__init__.py]: Actually, this you removed and didn't put back. Same for the other ones - that explains the CI failure

**Comment by ilayn** [comment]
@jakobjakobson13 Friendly nudge about online changes. You can still push your changes all at once at one go via git commits. Then you don't need to save things and trigger CI/CD pipelines with which we don't want to be throttled.

**Comment by jakobjakobson13** [comment]
> @jakobjakobson13 Friendly nudge about online changes. You can still push your changes all at once at one go via git commits. Then you don't need to save things and trigger CI/CD pipelines with which we don't want to be throttled.
> 
> If you need help with offline builds please ask for help.

Okay, got it.

**Comment by jakobjakobson13** [comment]
Before you push this comment to the master branch, I have the following comments:

- I also reduced to `constants.py` file by deleting all constants from that file. So this would break compatibility of scripts that used something like `scipy.constants.c` for the speed of light. Also the cleanup leaves only the auxiliary functions and so the filename gets misleading. I also deleted the prefixes even though I thought about putting them into a new file called `prefixes.py`.
- As the seven constants for the seven base units should have fixed values for the next years, I wanted to test them by hard coding their values to the test module. My local tests failed however as I could not test as the constants `hyperfine coupling constant of casium-133` could not be found.
- I would find a function that simplifies the units of a calculation quite convenient. To implement this however, you would have to convert all derived units into SI units (that would be easy), convert them to sympy variables (quite easy), simplify the resulting term (easy) and possibly reconvert them into a SI derived unit (quite tricky). But it would definitely be handy.

File: scipy/constants/__init__.py [hunk 5]
```diff
@@ -5,316 +5,118 @@
 
 .. currentmodule:: scipy.constants
 
-Physical and mathematical constants and units.
 
+About the SI units
```
Comment (pv) [inline scipy/constants/__init__.py:9]: This is reference documentation, so background material would better go to the bottom of the page, as people are mainly going to be looking for variable/function names etc. (except maybe the first time when reading this).

File: scipy/constants/constants.py [hunk 6]
```diff
@@ -1,178 +1,7 @@
...
-
-# force in newton
-dyn = dyne = 1e-5
-lbf = pound_force = pound * g
-kgf = kilogram_force = g  # * 1 kg
-
-# functions for conversions that are not linear
-
+from .codata import value
+import numpy as np
```
Comment (pv) [inline scipy/constants/constants.py:2]: This is no-go, these can't be just removed without breaking backward compat.

I'd recommend to stick with documentation changes in this PR, other unrelated changes should go to separate PRs.

**Review by pv — CHANGES_REQUESTED** [review]

File: scipy/constants/__init__.py (hunk 5 above)
Comment (jakobjakobson13) [reply scipy/constants/__init__.py:9]: Okay, I'll do it later.

File: scipy/constants/constants.py (hunk 6 above)
Comment (jakobjakobson13) [reply scipy/constants/constants.py:2]: > This is no-go, the constants in this file can't be just removed without breaking backward compat.

Okay, I'll revert it later.

> I'd recommend to stick with documentation changes in this PR, other unrelated changes should go to separate PRs.

Should I go for it anyway as it breaks backwards compatibility?

Comment (pv) [reply scipy/constants/constants.py:2]: > Should I go for it anyway as it breaks backwards compatibility?

I think not, as we can't remove them (and deprecation etc. does not seem warranted). But as a general comment, it's better to stick to one topic in one PR.

**Comment by jakobjakobson13** [comment]
As I deleted my scipy repository and the specific branch, I'm closing this pull request.

**Comment by tupui** [comment]
@jakobjakobson13 do you still plan to work on this?

**Comment by jakobjakobson13** [comment]
> @jakobjakobson13 do you still plan to work on this?

That´s up to you.

However, if I work on it, I´ll slim it down drastically with the main page consisting of the following information:
1. The seven SI defining constants of the SI [1]
2. The CODATA contants given in [2, 3] (as Scipy currently provides them).

And a subpage giving the following background:
1. The SI base units [1]
2. Base quantities and dimensions used in the SI [1]
3. SI prefixes [1]
4. The 22 SI units with special names and symbols [1]
5. Examples of coherent derived units in the SI expressed in terms of base
units [1]
6. Examples of SI coherent derived units whose names and symbols
include SI coherent derived units with special names and symbols [1]
7. Non-SI units accepted for use with the SI Units [1]

[1] [The International System of Units – 9th edition – Complete text in English and French (2019)](https://www.bipm.org/documents/20126/41483022/SI-Brochure-9.pdf/fcf090b2-04e6-88cc-1149-c3e029ad8232?version=1.18&t=1645193776058&download=true)
[1] [E. Tiesinga, P. J. Mohr, D. B. Newell, and B. N. Taylor, J. Phys. Chem. Ref. Data 50, 033105 (2021)](https://physics.nist.gov/cuu/pdf/JPCRD2018CODATA.pdf)
[2] [E. Tiesinga, P. J. Mohr, D. B. Newell, and B. N. Taylor, Rev. Mod. Phys. 93, 025010 (2021)](https://physics.nist.gov/cuu/pdf/RevModPhys.93.025010.pdf)

**Comment by jakobjakobson13** [comment]
Also, I deleted my fork on my hard drive and on Github. Does this mean, I would have to restart from scratch?

**Comment by tupui** [comment]
You seems to be the reference for this module so I would trust your judgment.

> Also, I deleted my fork on my hard drive and on Github. Does this mean, I would have to restart from scratch?

No 😃 The changes are still available here and you can use for instance GitHub's CLI to pull it with `gh pr checkout 11682`

**Comment by jakobjakobson13** [comment]
After second thoughts, I would defer this pull request to some time later. Feel free to take over.
